Add the major version of the release as suffix to the package name

andreas-bok-sociomantic opened the pull request:
Addresses problem where apt-get install a newer version over an older. 
The PR adds the major version as suffix to the libname. 
Addresses #149

mathias-lang-sociomantic commented:
Squash ?

andreas-bok-sociomantic commented:
Yes, but still a WIP, so will pingpong until the changes are more stable

leandro-lucarella-sociomantic commented on deb/build:
```diff
@@ -22,6 +22,38 @@ genchangelog()
+    ln -sf $so_libname.$libversion $so_libname.$lib_major
+    ln -sf $so_libname.$lib_major $so_libname
+
+    cd $curr_dir
```
No need to save this, you can access the previous directory with `$OLDPWD` and you can just use `cd -` as a shortcut to `cd $OLDPWD`.

leandro-lucarella-sociomantic commented on deb/build:
```diff
@@ -22,6 +22,38 @@ genchangelog()
+    # rename the original library file to the versioned library file
+    mv $pkgname.so.$libversion "$pkgname$pkgsuffix.so.$libversion"
+    mv $pkgname.a "$pkgname$pkgsuffix.a"
+    mv $pkgname.la "$pkgname$pkgsuffix.la"
```
Why not doing all this when creating the package via `fpm` mappings instead? It looks overly complicated.

leandro-lucarella-sociomantic commented on m4/ax_package_versioned.m4:
```diff
@@ -0,0 +1,30 @@
+# ===========================================================================
+# https://github.com/andreas-bok-sociomantic
```
Why not using the upstream project if you really need to specify an URL?

leandro-lucarella-sociomantic commented on m4/ax_package_versioned.m4:
```diff
@@ -0,0 +1,30 @@
+
+    major=`echo $VERSION | cut -f 1 -d .`
+    AC_SUBST([PACKAGE_VERSIONED],["$PACKAGE$major"])
+    ])
```
Also do you really need a m4 extension for this? Why not doing it in the `deb/build` script?

andreas-bok-sociomantic committed:
Define the library api level in configure.ac

The api level is exposed as API_LEVEL

Redefine PACKAGE_TARNAME to contain the api level

andreas-bok-sociomantic committed:
Sanitize deb/build script

- Remove obsolete variables
- Add documentation to the script
- Fix insane file test

andreas-bok-sociomantic committed:
Create deb packages with api level specified

Pass API_LEVEL defined in configure.ac as cmd
argument to deb/build

Refactor the deb/build script so the API_LEVEL is part of
the generated package name and library files

Include headers are mapped to a subdirectory of
/usr/install/libdrizzle-redux. Eg. for an api level
of 6 and install prefix /usr the path would be:

  /usr/include/libdrizzle-redux/libdrizzle-redux6/libdrizzle-redux

In effect include development headers the -I flag should be
-I$prefix/include/libdrizzle-redux/libdrizzle-redux6

andreas-bok-sociomantic commented:
Restructured based on comments from luca

andreas-bok-sociomantic commented:
> I don't think the headers should include it, for that it might be better to install headers in a different directory, i.e.: /usr/include/libdrizzle-redux/libdrizzle-redux6/libdrizzle-redux so then an application could use -I/usr/include/libdrizzle-redux/libdrizzle-redux6

The changes doesn't yet include the installation of include into a sub directory of `/usr/include/libdrizzle-redux/` as suggested by @leandro-lucarella-sociomantic in [issue 149](https://github.com/sociomantic-tsunami/libdrizzle-redux/issues/149#issuecomment-295786481)

andreas-bok-sociomantic commented:
Updated: 
- Fixed wrong declaration of `dist_pkgname`
- Added the mapping for the include headers to 

    `/usr/include/libdrizzle-redux/libdrizzle-redux6/libdrizzle-redux` 
```diff
diff --git a/deb/build b/deb/build
index 614a731..1270c0c 100755
--- a/deb/build
+++ b/deb/build
@@ -44,6 +44,12 @@ pkgname=libdrizzle-redux
 # library .so name
 so_libname=$pkgname.so
 
+# get the api level for the package from the cmd args
+api_level=""
+if [ $# -eq 1 -a ! -z "$1" ]; then
+    api_level=$1
+fi
+
 # the package name used in the distribution
 dist_pkgname=$pkgname$api_level
 
@@ -63,12 +69,6 @@ fi
 libversion=`echo $version_number | grep -o -P -e "([0-9]+\.?)+$"`
 lib_major=$(echo $libversion | cut -f1 -d .)
 
-# get the api level for the package from the cmd args
-api_level=""
-if [ $# -eq 1 -a ! -z "$1" ]; then
-    api_level=$1
-fi
-
 # create temp file to write changelog to
 changelog=`mktemp`
 trap "rm -f '$changelog'; exit 1" INT TERM QUIT
@@ -140,6 +140,6 @@ fpm -s dir -t deb -n "$dist_pkgname-dev" -v "$pkgversion" \
     --depends $dist_pkgname \
     --deb-changelog "$changelog" \
     --deb-no-default-config-files \
-    install/usr/include/=/usr/ \
+    install/usr/include/include/=/usr/include/$pkgname/$dist_pkgname/ \
     $libdir/$dist_so_libname=/usr/lib/ \
     $libdir/$pkgname.a=/usr/lib/$dist_pkgname.a

```

ben-palmer-sociomantic commented on deb/build:
```diff
@@ -44,15 +69,29 @@ fi
+cd $libdir
+touch $dist_so_libname.$libversion
+ln -sf $dist_so_libname.$libversion $dist_so_libname.$lib_major
+ln -sf $dist_so_libname.$lib_major $dist_so_libname
```
So this line will replace the current `libdrizzle-redux` with `libdrizzle-redux.6` right? And that is intended?

andreas-bok-sociomantic replied:
No the api level is appended right after the pkgname, ie. `api_level=6` would generate the package  `libdrizzle-redux6`

ben-palmer-sociomantic replied:
Ah ok, but this is the intended behaviour? Just wanted to confirm I understand whats going on.

andreas-bok-sociomantic replied:
Yeah. It would change the linking to -ldrizzle--redux6 for versions major branch 6.x.x.  Some libraries have use major-minor to distinguish the api level of the library, e.g. `somelib-2.0.so.1.0.1` but given that semver doesn't break ABI compatibility I reckon major version is sufficient.

ben-palmer-sociomantic commented on deb/build:
```diff
@@ -66,41 +105,41 @@ fpm -s dir -t deb -n "$pkgname" -v "$pkgversion" \
-    install/usr/include/=/usr/ \
-    $so_libname=/usr/lib/ \
-    install/usr/lib/libdrizzle-redux.a=/usr/lib/
+    install/usr/include/include/=/usr/include/$pkgname/$dist_pkgname/ \
```
Is `include/include` correct?

andreas-bok-sociomantic replied:
Yes, with the new project layout the make install create a two `include` subdirs.

ben-palmer-sociomantic reviewed (commented):
I had a couple of queries but generally LGTM.

andreas-bok-sociomantic committed:
Specify api level for RPM packages

Appends the api level to the generated rpm
package name, e.g:

    libdrizzle-redux6-6.0.1-1.x86_64.rpm

Change use of autorconf @PACKAGE@ variable to
@PACKAGE@@API_LEVEL@

andreas-bok-sociomantic commented:
Updated 
- Specify api level for RPM packages

```diff 
diff --git a/configure.ac b/configure.ac
index 63ec798..b19f3d4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -16,6 +16,12 @@ AC_CANONICAL_TARGET
 AM_INIT_AUTOMAKE(1.11 no-define color-tests -Wno-portability subdir-objects foreign tar-ustar m4_ifndef([AM_WITH_REGEX], [serial-tests]))
 AC_PREREQ([2.68])
 
+# Define the api level of the package.
+AC_SUBST([API_LEVEL],[6])
+
+# Add the api level to the name of the generated tar file
+AC_SUBST([PACKAGE_TARNAME],[$PACKAGE_NAME$API_LEVEL])
+
 AC_CANONICAL_HOST
 AC_CANONICAL_BUILD
 AC_ARG_PROGRAM
@@ -144,8 +150,6 @@ AX_APPEND_COMPILE_FLAGS([-I./include], [CFLAGS])
 AX_CREATE_GENERIC_CONFIG
 AX_AM_JOBSERVER([yes])
 
-AC_SUBST([API_LEVEL],[6])
-
 AC_CONFIG_FILES([Makefile
                  docs/conf.py
                  rpm/spec
diff --git a/rpm/include.am b/rpm/include.am
index 4e76906..7b78dc1 100644
--- a/rpm/include.am
+++ b/rpm/include.am
@@ -2,8 +2,8 @@
 
 rpm-build: rpm/spec dist
 	@rm -f *.rpm
-	@rm -f ~/rpmbuild/RPMS/x86_64/$(PACKAGE)-$(VERSION)*.rpm
-	@rm -f ~/rpmbuild/SRPMS/$(PACKAGE)-$(VERSION)*.rpm
+	@rm -f ~/rpmbuild/RPMS/x86_64/$(PACKAGE)$(API_LEVEL)-$(VERSION)*.rpm
+	@rm -f ~/rpmbuild/SRPMS/$(PACKAGE)$(API_LEVEL)-$(VERSION)*.rpm
 	@mkdir -p ~/rpmbuild/BUILD/
 	@mkdir -p ~/rpmbuild/RPMS/i386/
 	@mkdir -p ~/rpmbuild/RPMS/i686/
@@ -12,11 +12,11 @@ rpm-build: rpm/spec dist
 	@mkdir -p ~/rpmbuild/SOURCES/
 	@mkdir -p ~/rpmbuild/SPECS/
 	@mkdir -p ~/rpmbuild/SRPMS/
-	@cp $(PACKAGE)-$(VERSION).tar.gz ~/rpmbuild/SOURCES/
+	@cp $(PACKAGE)$(API_LEVEL)-$(VERSION).tar.gz ~/rpmbuild/SOURCES/
 	@rpmbuild -ba --clean rpm/spec
-	@cp ~/rpmbuild/RPMS/x86_64/$(PACKAGE)-$(VERSION)*.rpm .
-	@cp ~/rpmbuild/RPMS/x86_64/$(PACKAGE)-devel-$(VERSION)*.rpm .
-	@cp ~/rpmbuild/SRPMS/$(PACKAGE)-$(VERSION)*.rpm .
+	@cp ~/rpmbuild/RPMS/x86_64/$(PACKAGE)$(API_LEVEL)-$(VERSION)*.rpm .
+	@cp ~/rpmbuild/RPMS/x86_64/$(PACKAGE)$(API_LEVEL)-devel-$(VERSION)*.rpm .
+	@cp ~/rpmbuild/SRPMS/$(PACKAGE)$(API_LEVEL)-$(VERSION)*.rpm .
 
 rpm-sign: rpm-build
 	@rpm --addsign *.rpm
diff --git a/rpm/spec.in b/rpm/spec.in
index 71672cc..f17ec79 100644
--- a/rpm/spec.in
+++ b/rpm/spec.in
@@ -1,5 +1,5 @@
 Summary: Simplified API to MySQL databases
-Name: @PACKAGE@
+Name: @PACKAGE@@API_LEVEL@
 Version: @VERSION@
 Release: 1
 License: BSD
@@ -11,7 +11,7 @@ URL: https://github.com/sociomantic-tsunami/libdrizzle-redux
 %define packager_email %(git config user.email)
 Packager: %{packager_name} <%{packager_email}>
 
-Source: libdrizzle-redux-%{version}.tar.gz
+Source: %{name}-%{version}.tar.gz
 BuildRoot: %{_tmppath}/%{name}-%{version}-%{release}-buildroot
 
 %description

```

ben-palmer-sociomantic commented:
LGTM. Do you want any more reviews or are you happy for me to rebase?

andreas-bok-sociomantic commented:
> LGTM. Do you want any more reviews or are you happy for me to rebase?

I might just wanna see if @leandro-lucarella-sociomantic has any comments. He after all has a `deb-conf` T-Shirt :)

leandro-lucarella-sociomantic commented:
> He after all has a deb-conf

It is *DebConf* and I was just visiting some friends :P

leandro-lucarella-sociomantic commented on deb/build:
```diff
@@ -2,6 +2,14 @@
+#
+# The `api_level` will be appended to the `package_name` for the generated
+# packages, i.e. running `./build liblicious 6` would generate a packages named
+# `liblicious6`, `liblicious6-dbg`, `liblicious6-dev`.
```
Why liblicious?

andreas-bok-sociomantic replied:
Bad joke, sorry

leandro-lucarella-sociomantic commented on deb/build:
```diff
@@ -32,10 +40,27 @@ pkgmaint=$(echo "`git config user.name` <`git config user.email`>")
+
+# get the ABI library version from the generated .so file
+version_number=`find $libdir ! -type l -name "$so_libname.*"`
 if [ $(echo $version_number | wc -l) != 1 ]; then
```
Not part of this PR, but this is a very weird way to check if a file is found or not :thinking:

andreas-bok-sociomantic replied:
Hehe, saying that `-e` is easier...

andreas-bok-sociomantic replied:
It could meaningfully go into the commit "Sanitize deb/build script"

leandro-lucarella-sociomantic reviewed (commented):
I did just a very quick pass, honestly I'm not an expert in regards of shared library versioning or autotools :/

andreas-bok-sociomantic commented:
Updated based on comments from @leandro-lucarella-sociomantic 

```diff
diff --git a/Makefile.am b/Makefile.am
index 89864fe..5a274c9 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -11,6 +11,7 @@ CLEANFILES =
 DISTCLEANFILES =
 bin_PROGRAMS =
 dist_bin_SCRIPTS =
+distdir = @PACKAGE@@API_LEVEL@-@VERSION@
 noinst_HEADERS =
 lib_LTLIBRARIES =
 man_MANS =
@@ -69,8 +70,8 @@ maintainer-clean-local:
 	-rm m4/ltversion.m4
 	-rm m4/lt~obsolete.m4
 	find . -type f -name '*~' -exec rm -f '{}' \;
-	-rm -f @PACKAGE@-*.tar.gz
-	-rm -f rpm/@PACKAGE@-*.rpm
-	-rm -f @PACKAGE@-*.deb
+	-rm -f @PACKAGE@@API_LEVEL@-*.tar.gz
+	-rm -f rpm/@PACKAGE@@API_LEVEL@-*.rpm
+	-rm -f @PACKAGE@@API_LEVEL@-*.deb
 
 dist_bin_SCRIPTS+= @GENERIC_CONFIG@
diff --git a/deb/build b/deb/build
index 1270c0c..5ccb01b 100755
--- a/deb/build
+++ b/deb/build
@@ -8,8 +8,8 @@
 #     api_level        the api level for the package
 #
 # The `api_level` will be appended to the `package_name` for the generated
-# packages, i.e. running `./build liblicious 6` would generate a packages named
-# `liblicious6`, `liblicious6-dbg`, `liblicious6-dev`.
+# packages, i.e. running `./build 6` would generate a packages named
+# `libdrizzle-redux6`, `libdrizzle-redux-dbg`, `libdrizzle-redux-dev`.
 
 set -e
 
@@ -61,7 +61,7 @@ libdir="./install/usr/lib"
 
 # get the ABI library version from the generated .so file
 version_number=`find $libdir ! -type l -name "$so_libname.*"`
-if [ $(echo $version_number | wc -l) != 1 ]; then
+if [ ! -e $version_number ]; then
     echo "'.so' file not found"
     exit
```

ben-palmer-sociomantic commented on deb/build:
```diff
@@ -1,8 +1,14 @@
+
 set -e
 
+# change directory to containing deb `build` script
```
Perhaps better as `change directory to the deb `build` script directory`

ben-palmer-sociomantic commented on deb/build:
```diff
@@ -2,6 +2,14 @@
+#
+# The `api_level` will be appended to the `package_name` for the generated
+# packages, i.e. running `./build 6` would generate a packages named
+# `libdrizzle-redux6`, `libdrizzle-redux-dbg`, `libdrizzle-redux-dev`.
```
So `libdrizzle-redux6` is the only lib with the suffix? Not `libdrizzle-redux-dbg` and `libdrizzle-redux-dev`?

ben-palmer-sociomantic reviewed (commented):
Couple of small comments then LGTM.

andreas-bok-sociomantic commented:
Updated based on comments from @ben-palmer-sociomantic 
```diff
diff --git a/deb/build b/deb/build
index a117446..a0c909a 100755
--- a/deb/build
+++ b/deb/build
@@ -9,11 +9,11 @@
 #
 # The `api_level` will be appended to the `package_name` for the generated
 # packages, i.e. running `./build 6` would generate a packages named
-# `libdrizzle-redux6`, `libdrizzle-redux-dbg`, `libdrizzle-redux-dev`.
+# `libdrizzle-redux6`, `libdrizzle-redux6-dbg`, `libdrizzle-redux6-dev`.
 
 set -e
 
-# change directory to containing deb `build` script
+# cd to the directory containing the deb `build` script
 cd `dirname $0`
 
 # function to automatically generate changelog
```

andreas-bok-sociomantic committed:
Update section in README and rst docs about linking

Adds description of how to link against
libdrizzle-redux for versions >=v6.0.2

andreas-bok-sociomantic commented:
Updated
- Rebased on v6.0.x
- Fixed typo
- Updated sections about linking in README.rst and compiling.rst

ben-palmer-sociomantic commented on README.rst:
```diff
@@ -71,10 +71,19 @@ refer to `compiling.rst`_ or build the documentation, cf. below.
 **Linking libdrizzle-redux**::
 
+Ensure the library is in your library and include paths. For releases prior to
+`v6.0.3` linking your app against libdrizzle-redux requires ``-ldrizzle-redux``::
```
`v6.0.3` or `v6.0.2`?

andreas-bok-sociomantic replied:
Ah, damn it. `v6.0.2`

ben-palmer-sociomantic commented:
One comment, then LGTM.

andreas-bok-sociomantic commented:
Fixed wrong version number in last commit

ben-palmer-sociomantic commented:
LGTM

andreas-bok-sociomantic commented:
Tx
